### PR TITLE
Update HQ's JS Tests using mocha-chrome-headless

### PR DESCRIPTION
**Pull request by biyeun**
https://dimagi-dev.atlassian.net/browse/SAAS-10973

##### SUMMARY
I hope travis doesn't bork on this...but creating this PR to test it out. 🤞 

You can now run JS tests using:
```
grunt test
```

or a specific test, like `app_manager`, with
```
grunt test:app_manager
```

It's recommended you blow away your `node_modules` folder and do a fresh `npm install` in order to access these new tests (once merged in).

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -1,77 +1,122 @@
...
-            };
-        };
+        grunt.log.writeln("\n");
+        grunt.log.writeln(testText.bold);
+        grunt.log.write(currentTestPath.italic.cyan);
+
+        headless.runner({
+            file: currentTestPath,
+            visible: false,
+            timeout: 120000,
```
Comment (orangejenny) [inline Gruntfile.js]: Does this mean there's a 2-minute timeout on each individual test?

File: Gruntfile.js [hunk 2]
```diff
@@ -1,77 +1,122 @@
...
+    grunt.task.registerTask(
+        'test',
+        'Runs Javascript Tests. Pass in an argument to run a specific test',
+        function (arg) {
+            var paths = _.concat(apps, custom);
+            if (arg) {
+                paths = [arg];
+            }
+            var testStatement = "Running tests: " + paths.join(', ');
+            grunt.log.writeln(testStatement.bold.green);
```
Comment (orangejenny) [inline Gruntfile.js:121]: This `content.style.style` syntax is interesting.

File: docker/run.sh [hunk 3]
```diff
@@ -32,7 +32,9 @@ function setup() {
     fi
 
     if [ "$TEST" = "javascript" -o "$JS_SETUP" = "yes" ]; then
-        npm install npm@latest
+        curl -sL https://deb.nodesource.com/setup | sudo bash - && \
+        apt-get install -yq nodejs build-essential
+        npm install -g npm
```
Comment (orangejenny) [inline docker/run.sh]: Is the next line still needed? Seems like this is doing a global install then a local one, is that accurate?

File: Gruntfile.js [hunk 4]
```diff
@@ -59,18 +59,24 @@ module.exports = function(grunt) {
...
+            testText = "Running Test '" + currentApp + "'",
+            runner_options = {
+                file: currentTestPath,
+                visible: false,
+                timeout: 120000,
+                reporter: 'dot',
+            };
+
+        // For running in docker/travis
+        if (process.env.PUPPETEER_SKIP_CHROMIUM_DOWNLOAD) {
```
Comment (orangejenny) [inline Gruntfile.js:71]: What's puppeteer?

**Review by orangejenny — APPROVED** [review]
This is fabulous!

File: docker/run.sh (hunk 3 above)
Comment (orangejenny) [reply docker/run.sh]: nevermind, this was updated in a later commit

**Review by calellowitz — APPROVED** [review]
Having zero context on JS tests, looks fine to me, and great that we are getting rid of `grunt-mocha`

File: DEV_SETUP.md [hunk 5]
```diff
@@ -490,12 +490,6 @@ Occasionally you will see an app specified with a `#`, like `app_manager#b3`. Th
 http://localhost:8000/mocha/<app_name>/<config>  // (e.g. http://localhost:8000/mocha/app_manager/b3)
 ```
 
-### Continuous JavaScript testing
-
-By running the `watch` command, it's possible to continuously run the JavaScript test suite while developing
-
-    $ grunt watch:<app_name>  // (e.g. grunt watch:app_manager)
```
Comment (millerdev) [inline DEV_SETUP.md:497]: I thought watch was a feature of mocha. Is this not possible anymore?

File: Gruntfile.js [hunk 6]
```diff
@@ -1,77 +1,128 @@
 module.exports = function(grunt) {
+    var headless = require('mocha-headless-chrome'),
+        _ = require('lodash'),
```
Comment (millerdev) [inline Gruntfile.js:3]: Does this mean we now depend on both underscore and lodash?

**Review by millerdev — APPROVED** [review]
:100: Nice!

File: DEV_SETUP.md (hunk 5 above)
Comment (biyeun) [reply DEV_SETUP.md:497]: not with the way the current integration works

File: Gruntfile.js (hunk 6 above)
Comment (biyeun) [reply Gruntfile.js:3]: lodash was already installed as a dependency from one of the other packages, so I just ended up using it and requiring it in `package.json`. Arguably, we should move to `lodash` in favor of ditching `underscore`, but right now `npm`'s `lodash` and `bower`'s `underscore` are in completely different places / different use cases

File: Gruntfile.js (hunk 4 above)
Comment (biyeun) [reply Gruntfile.js:71]: `puppeteer` is google's response to `selenium` for running ui tests in chrome I believe, but `mocha-chrome-headless` requires it as a dependency.

**Pull request merged**